### ISSUE TITLE
Question for CutMix in loading data

### ISSUE THREAD

**Issue by ijnokml** [LiheYoung/UniMatch#51]
Hello Young,

Thank you for your exceptional work. I've rerun the project, and the performance is indeed impressive.

However, I have some questions about the cutmix part in the file `unimatch.py` in line 142-145. It seems that the cutmix isn't working. I try to compare the array of `img_u_s1` before the line 142 and after the line 145, and I found that they are seems the same array. As I known, the same dataloader loaded the batch with the same data in a iter. 
I am still learning about this paper, and I think I might misunderstand the code. Could the you provide some help or guides?

 The code in the file `unimatch.py` in line 142-145 are listed below:
```
            img_u_s1[cutmix_box1.unsqueeze(1).expand(img_u_s1.shape) == 1] = \
                img_u_s1_mix[cutmix_box1.unsqueeze(1).expand(img_u_s1.shape) == 1]
            img_u_s2[cutmix_box2.unsqueeze(1).expand(img_u_s2.shape) == 1] = \
                img_u_s2_mix[cutmix_box2.unsqueeze(1).expand(img_u_s2.shape) == 1]
```

**Comment by LiheYoung** [comment]
Hi,

The ``img_u_s1`` and ``img_u_s1_mix`` come from the same dataloader, and the share the same image index (the same raw image). **However, they are indeed different views of the same raw image**, because we apply random weak augmentations (resize and crop) and random strong augmentation on them. Therefore, performing CutMix on them is also beneficial.

Of course, a more natural choice is to mix two images from two different raw images. Honestly, we previously wrongly thought ``img_u_s1`` and ``img_u_s1_mix`` are of different image indexes. After realizing they share the same raw image, we tried to use two different image indexes for CutMix, but did not observe any improvement (comparable results). So you can use our current code for CutMix, or use a more natural and correct way of mixing two images from two different raw images.

**Comment by ijnokml** [comment]
Thanks for your reply, your explanation has been very helpful in understanding the code. However, based on my experiments, since the `img_u_s1` and `img_u_s1_mix` come from the same dataloader object, they seem to have the same augmentation (here I subtract the array before and after CutMix, and I always get 0). I thought perhaps using a new `trainloader_u_copy` as a supplement might solve the problem? Would you be willing to share the code for "use two different image indexes for CutMix" with me? So that I can verify the effectiveness of CutMix. In any case, I think your work is elegant and effective, and thank you once again for your reply.

**Comment by LiheYoung** [comment]
You can try ``img_u_s1 - img_u_s1_mix``, and will find their difference.

For the code of **using two different image indexes for CutMix**, you can simply create an auxiliary  ``dataloader_u_mix``:
```python
trainsampler_u_mix = torch.utils.data.distributed.DistributedSampler(trainset_u)
trainloader_u_mix = DataLoader(trainset_u, batch_size=cfg['batch_size'],
                               pin_memory=True, num_workers=1, drop_last=True, sampler=trainsampler_u_mix)
```

Then, provide different shuffling seeds for ``dataloader_u_mix`` and the original ``dataloader_u`` (L119 of [unimatch.py](https://github.com/LiheYoung/UniMatch/blob/main/unimatch.py)):
```python
trainloader_u.sampler.set_epoch(epoch)
trainloader_u_mix.sampler.set_epoch(epoch+cfg['epochs'])

loader = zip(trainloader_l, trainloader_u, trainloader_u_mix)
```
Others are kept the same.

**Comment by ijnokml** [comment]
Thank you for your reply. I tried the `dataloader_u_mix` and I got below 1% improve on some metrics. The changes works, but they seem not important to the whole work. Thank you on your explaining on the `CutMix`.
